hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,197 @@
+
+#ifndef MPCD_SPHERE_GEOMETRY_H_
+#define MPCD_SPHERE_GEOMETRY_H_
+
+#include "BoundaryCondition.h"
+
+#include "hoomd/HOOMDMath.h"
+#include "hoomd/BoxDim.h"
+
+#include <cmath>
+
+#ifdef NVCC
+#define HOSTDEVICE __host__ __device__ inline
+#else
+#define HOSTDEVICE inline __attribute__((always_inline))
+#include <string>
+#endif // NVCC
+
+namespace mpcd
+{
+namespace detail
+{
+//! Spherical droplet geometry
+/*!
+ * TODO: write proper description of what is being done in the class below
+ */
+class __attribute__((visibility("default"))) SphereGeometry
+    {
+    public:
+        //! Constructor
+        /*!
+         * \param R confinement radius
+         * \param bc Boundary condition at the wall (slip or no-slip)
+         */
+        HOSTDEVICE SphereGeometry(Scalar R, boundary bc)
+            : m_R(R), m_R2(R*R), m_bc(bc)
+            { }
+
+        //! Detect collision between the particle and the boundary
+        /*!
+         * \param pos Proposed particle position
+         * \param vel Proposed particle velocity
+         * \param dt Integration time remaining
+         *
+         * \returns True if a collision occurred, and false otherwise
+         *
+         * \post The particle position \a pos is moved to the point of reflection, the velocity \a vel is updated
+         *       according to the appropriate bounce back rule, and the integration time \a dt is decreased to the
+         *       amount of time remaining.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+
+            /*
+             * Detect if particle has left the spherical confinement.
+             * comparison: 'r2 <= R*R' used in the following calculations is equal to 0(FALSE) if the particle is outside
+             * a sphere of radius R, 1(TRUE) if inside.
+             */
+            Scalar r2 = dot(pos,pos);
+            // exit immediately if no collision is found.
+            /*
+            To avoid division by zero error later on, we exit immediately. (no collision could have occurred
+            if the particle speed is equal to zero)
+            */
+            Scalar v2 = dot(vel,vel);
```

Comment:
Condense this comment, and use `const` when the variable will not be changed later
```suggestion
            // If particle is still inside sphere or has zero speed, no collision occurred.
            const Scalar r2 = dot(pos,pos);
            const Scalar v2 = dot(vel,vel);
```